File: lintcode/Longest_Increasing_Subsequence.cpp
class Solution {
public:
    /**
     * @param nums: The integer array
     * @return: The length of LIS (longest increasing subsequence)
     */
    int longestIncreasingSubsequence(vector<int> nums) {
        // write your code here
        if(nums.size() <= 1) return nums.size();
        
        int n = nums.size();
        vector<int> L(n, 1);
        int maxLen = 1;
        
        for(int i = 1; i < n; i++){
            for(int j = 0; j < i; j++){
                int t;
                if(nums[i] >= nums[j]) t = L[j] + 1;
                else t = 1;
                
                if(t > L[i]) L[i] = t;
            }
            if(L[i] > maxLen) maxLen = L[i];
        }
        
        return maxLen;
    }
};

File: lintcode/java/LongestIncreasingSubsequence.java
public class Solution {
    /**
     * @param nums: The integer array
     * @return: The length of LIS (longest increasing subsequence)
     */
    public int longestIncreasingSubsequence(int[] nums) {
        if(nums.length == 0) return 0;
        
        int n = nums.length;
        // t[i]: LIS ending with nums[i]
        int[] t = new int[n];
        Arrays.fill(t, 1);
        
        int maxLen = 1;
        for(int i = 1; i < n; i++){
            for(int j = 0; j < i; j++){
                if(nums[i] >= nums[j]){
                    t[i] = Math.max(t[j]+1, t[i]);
                }
            }
            if(t[i] > maxLen){
                maxLen = t[i];
            }
        }
        return maxLen;
    }
}

